Add metrics from /metrics/snapshot endpoint for master and slaves

pvdissel opened the pull request:
To gather more in-depth metrics, metrics from the default provided
`/metrics/snapshot` endpoint is now also collected. This is based
on the `/metrics/snapshot` response of Mesos 0.22.1

Master is just as important as the slaves, therefor now collecting
the master and slave metrics from the `/metrics/snapshot` endpoint.

The first scrape was only done after the first interval, this
is now fixed to also do an initial scrape on startup.

ps. This is my first-time with Go, so there are probably things that could be done cleaner. But it works :smile:

brian-brazil commented on main.go:
```diff
@@ -14,50 +14,21 @@ import (
 	"github.com/antonlindstrom/mesos_stats"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/log"
+	"github.com/pvdissel/mesos_metrics"
```
This library lacks license information, we usually use Apache 2

pvdissel replied:
Will switch to Apache2 licensed one

brian-brazil commented on main.go:
```diff
@@ -14,50 +14,21 @@ import (
+	addr = flag.String("web.listen-address", ":9105", "Address to listen on for web interface and telemetry")
+	autoDiscover = flag.Bool("exporter.discovery", false, "Discover all Mesos slaves")
+	localURL = flag.String("exporter.local-url", "http://127.0.0.1:5051", "URL to the local Mesos slave")
+	masterURL = flag.String("exporter.discovery.master-url", "http://mesos-master.example.com:5050", "Mesos master URL")
```
Doe this allow running one of these beside each master? That's the preferred way to run exporters as it avoids a spof, spreads out load and keeps the SD logic in one place.

pvdissel replied:
It works just as it did before.
If you use the exporter.discovery option with exporter.discovery.master-url, it will ask the elected master for its slaves and will scrape that master and its slaves.
If you provide only the exporter.local-url it will scrape only that specific node as slave.

If you run an exporter for each master, that will work fine, as it will automatically go to the elected master.
So yes, this will allow running one of these beside each master :smile:

brian-brazil replied:
So if I run one of these beside each master and scrape each of them, then I'll get only data from that master?

pvdissel replied:
No, you'll get the data from the complete cluster based on the elected master of that cluster. But all masters should be in-sync with the elected master. Or... is that exactly what you want to know.. hmm.. good point.
In the original code it already went to the elected master and started from there, but maybe good to not go to the elected master but start at that specific master that's provided.
Not sure how that would work in the metrics gathering, as you're actually fetching the same info from different places. And there may-be a difference in values because of different times of scraping and possibly not completely in-sync masters.

What do you suggest?

brian-brazil replied:
Masters are always a bit messy due to being multiple of them that are (in-theory) in sync. If it's a situation where there's an elected leader that is authoritative for long periods of time (hours) and handles all traffic then you have the option of what you're doing here - but if the other masters do anything that's worth monitoring then you also need to scrape them.

What you end you usually doing is have some exporter sitting beside each master that scrapes only to that master. Depending on whether the master is leader or follower it may return different metrics, and then you handle all that on the prometheus end (via averaging of some form usually and/or only aggregating up stats from masters identifying themselves as the leader).

Slaves are simpler, you run an exporter beside each that prometheus scrapes. That takes care of the scrape timing and timestamps.

pvdissel replied:
Ok, thanks for the explanation. So it would be good to also be able to only scrape a master, I'll work on that

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,266 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.counter(metricsChan, labels, labelValues, "mesos_valid_framework_messages", float64(stats.ValidFrameworkMessages), "Valid framework messages")
+	e.counter(metricsChan, labels, labelValues, "mesos_valid_status_updates", float64(stats.ValidStatusUpdates), "Valid status updates")
+
+	e.gauge(metricsChan, labels, labelValues, "mesos_system_cpus_total", float64(stats.SystemCpuTotal), "System CPU total")
```
The node exporter already has these, no need to duplicate it.

pvdissel replied:
Ah, thanks for the heads-up. Haven't seen https://github.com/prometheus/node_exporter before, will check that out also

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,266 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.gauge(metricsChan, labels, labelValues, "mesos_cpus_total", float64(stats.CpuTotal), "CPU total")
+	e.gauge(metricsChan, labels, labelValues, "mesos_cpus_used", stats.CpuUsed, "CPU used")
+	e.gauge(metricsChan, labels, labelValues, "mesos_disk_percent", stats.DiskPercentage, "Disk percentage")
+	e.gauge(metricsChan, labels, labelValues, "mesos_disk_total", float64(stats.DiskTotal), "Disk total")
```
Only counters should end in `_total`. What is this metric?

pvdissel replied:
mesos_cpus_total is the total cpu amount that Mesos can offer as resource. In the master metrics this is what's in total available to the cluster (addition of all slave values). In the slave metrics this is what's available to the slave.

mesos_disk_total is the same as mesos_cpu_total, but then for total disk space.
I made it a gauge, as these values can fluctuate

brian-brazil replied:
something like `mesos_resource_disk_available_bytes` might be clearer

pvdissel replied:
Good one, will adjust it. Until now I just basicly 1-on-1 copied the keys from the Mesos metrics

brian-brazil replied:
There's always a tradeoff on whether to do a straight copy and keep maintenance simple, or to maintain it all by hand and get more useful metrics. Considering you're already listing each individual metric you may as well name them more usefully.

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,266 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.counter(metricsChan, labels, labelValues, "mesos_invalid_framework_messages", float64(stats.InvalidFrameworkMessages), "Invalid framework messages")
+	e.counter(metricsChan, labels, labelValues, "mesos_invalid_status_updates", float64(stats.InvalidStatusUpdates), "Invalid status updates")
+
+	e.gauge(metricsChan, labels, labelValues, "mesos_mem_percent", stats.MemoryPercentage, "Memory percentage")
```
Percentage what?

I think the below two metrics already cover this

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,266 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+		e.errors.WithLabelValues(host).Inc()
+		return
+	}
+	labels := []string{"node", "type"}
```
For completely different types of nodes you want a different job label.

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,266 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	}
+	labels := []string{"node", "type"}
+	labelValues := []string{host, "master"}
+	e.gauge(metricsChan, labels, labelValues, "mesos_cpus_percent", stats.CpuPercentage, "CPU percentage")
```
these should probably all be prefixed with mesos_master to avoid confusion with the slave stats

pvdissel replied:
aren't the labels for that? Next to first timer with Go, I also just started with Prometheus :smile:

So do I understand it correctly, that:
- Because master and slave nodes have a different purpose, their metric keys should be different and eg. contain "_master_" and "_slave_" respectively?
- The labels can be used to seperate metrics for different instances or call types?
  eg. for this situation having a "node" label with the hostname as value works nicely, because this indicates for which instance the metric is.
  And for something like `http_api_error_count` could have a label based on the specific http statuscode the error happened in?

Do I understand that correctly?

brian-brazil replied:
The metrics have different meanings, so they should have different names.

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,266 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.gauge(metricsChan, labels, labelValues, "mesos_disk_used", float64(stats.DiskUsed), "Disk used")
+	e.counter(metricsChan, labels, labelValues, "mesos_dropped_messages", float64(stats.DroppedMessages), "Dropped messages")
+
+	e.gauge(metricsChan, labels, labelValues, "mesos_elected", stats.DiskPercentage, "Elected")
```
DiskPercentage for elected?

pvdissel replied:
ah, good find. Will fix that

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,266 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.gauge(metricsChan, labels, labelValues, "mesos_cpus_percent", stats.CpuPercentage, "CPU percentage")
+	e.gauge(metricsChan, labels, labelValues, "mesos_cpus_total", float64(stats.CpuTotal), "CPU total")
+	e.gauge(metricsChan, labels, labelValues, "mesos_cpus_used", stats.CpuUsed, "CPU used")
+	e.gauge(metricsChan, labels, labelValues, "mesos_disk_percent", stats.DiskPercentage, "Disk percentage")
```
Is disk useful for a master? A master doesn't have resources to offer.

pvdissel replied:
The master has the total amount of the whole cluster as value.
So `mesos_cpus_total` on the master is the `mesos_cpus_total` added from all slaves.
And `mesos_disk_total` on the master is the `mesos_disk_total` added from all slaves.
etc.

brian-brazil replied:
Those need to be named differently then, something along the lines of `mesos_cluster_disk_available_bytes`.

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,266 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.gauge(metricsChan, labels, labelValues, "mesos_disk_percent", stats.DiskPercentage, "Disk percentage")
+	e.gauge(metricsChan, labels, labelValues, "mesos_disk_total", float64(stats.DiskTotal), "Disk total")
+	e.gauge(metricsChan, labels, labelValues, "mesos_disk_used", float64(stats.DiskUsed), "Disk used")
+	e.counter(metricsChan, labels, labelValues, "mesos_dropped_messages", float64(stats.DroppedMessages), "Dropped messages")
```
Counters should end with `_total`

brian-brazil commented:
I'm wondering if this should be two separate exporters, as you're monitoring two very different types of job.

pvdissel commented:
I believe I've adjusted all your feedback items, can you take a look.
FYI, things changed:
- Switched to slave, master and cluster specific metrics key names
- Renamed metrics to give them more context and follow the http://prometheus.io/docs/practices/naming/ best practices
- Removed all metrics that are also provided by the node_exporter. Except the NetworkStatistics as those were the original metrics, probably someone is depending on them
- Added some info on the Mesos endpoints used by the exporter and link to the node_exporter

brian-brazil commented on README.md:
```diff
@@ -26,10 +37,10 @@ mesos_exporter can operate in two modes: discovery and local.
 all slave metrics and export them. 
-This mode lets you have one exporter per Mesos cluster.
-
+This mode lets you have one exporter per Mesos cluster. But you can also have one exporter for each Mesos master
```
Do non-elected masters also expose `/metrics/snapshot`?

pvdissel replied:
Yes, they do. Every Mesos master and slave exposes it

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,251 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	labelValues := []string{host}
+	e.gauge(metricsChan, labels, labelValues, "mesos_slave_resource_cpus_available", float64(stats.CpuTotal), "Amount of CPUs available as resource")
+	e.gauge(metricsChan, labels, labelValues, "mesos_slave_resource_cpus_used", stats.CpuUsed, "Amount of CPU resources in use")
+	e.gauge(metricsChan, labels, labelValues, "mesos_slave_resource_cpus_used_percentage", stats.CpuPercentage, "Percentage of CPU resources in use")
```
If you're going to export this it should be as a ratio, not a percentage. Having a mix of ratios and percentages in a system gets confusing fast.

As this can be calculated from the other two, I'd suggest dropping it. This also reduces the risk of someone taking an average of it.

pvdissel replied:
Ok, will drop it

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,251 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.gauge(metricsChan, labels, labelValues, "mesos_slave_tasks_staging", float64(stats.TasksStaging), "Amount of staging tasks")
+	e.gauge(metricsChan, labels, labelValues, "mesos_slave_tasks_starting", float64(stats.TasksStarting), "Amount of starting tasks")
+
+	e.counter(metricsChan, labels, labelValues, "mesos_slave_uptime_seconds", stats.UptimeSecs, "Slave uptime in seconds")
```
I don't suppose there's a start time available rather than an uptime? If uptime gets stuck, you won't notice.

pvdissel replied:
Sadly no, as far as I could find. But I can transform the uptime to a start time (when I find the correct golang functions :smile:

brian-brazil replied:
There's no point in doing the transform, especially as it'd fluctuate due to races.

pvdissel replied:
hmm.. ok. So would a counter be a good metric type for an uptime value?

brian-brazil replied:
There's arguments both ways, I'd use a counter

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,251 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.counter(metricsChan, labels, labelValues, "mesos_master_event_queue_messages_total", float64(stats.EventQueueMessages), "Total of event queue messages")
+
+	e.gauge(metricsChan, labels, labelValues, "mesos_master_frameworks_active", float64(stats.FrameworksActive), "Amount of active frameworks")
+	e.counter(metricsChan, labels, labelValues, "mesos_master_frameworks_connected_total", float64(stats.FrameworksConnected), "Total of frameworks that have been connected")
```
I suspect this is more "Total frameworks connections that have occurred"

pvdissel replied:
:+1:

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,251 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.counter(metricsChan, labels, labelValues, "mesos_master_invalid_status_update_acknowledgements_total", float64(stats.InvalidStatusUpdateAcknowledgements), "Total of invalid status update acknowledgements")
+	e.counter(metricsChan, labels, labelValues, "mesos_master_invalid_status_updates_total", float64(stats.InvalidStatusUpdates), "Total of invalid status updates")
+
+	e.counter(metricsChan, labels, labelValues, "mesos_master_messages_authenticate_total", float64(stats.MessagesAuthenticate), "Total of authenticate messages")
```
Might all these be better as a single metric with labels for the message type?

pvdissel replied:
Like this? Or is there a better key name for this?

```
mesos_master_messages_total{type="authenticate"}
mesos_master_messages_total{type="deactivate_framework"}
mesos_master_messages_total{type="kill_task"}
```

brian-brazil replied:
'type' should be avoided as a label name, as it's very generic. 'message' would work in this case I think

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,251 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+
+	e.counter(metricsChan, labels, labelValues, "mesos_master_task_failed_source_slave_reason_command_executor_failed_total", float64(stats.TaskFailedSourceSlaveReasonCommandExecutorFailed), "Total tasks failed with as reason the command executor failed")
+
+	e.counter(metricsChan, labels, labelValues, "mesos_master_tasks_error_total", float64(stats.TasksError), "Total tasks in error")
```
These feel like a single metric too

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,251 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.counter(metricsChan, labels, labelValues, "mesos_master_tasks_finished_total", float64(stats.TasksFinished), "Total of finished tasks")
+	e.counter(metricsChan, labels, labelValues, "mesos_master_tasks_killed_total", float64(stats.TasksKilled), "Total of killed tasks")
+	e.counter(metricsChan, labels, labelValues, "mesos_master_tasks_lost_total", float64(stats.TasksLost), "Total of lost tasks")
+	e.gauge(metricsChan, labels, labelValues, "mesos_master_tasks_running", float64(stats.TasksRunning), "Amount of running tasks")
```
These 3 feel like a single metric

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,251 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.counter(metricsChan, labels, labelValues, "mesos_master_valid_status_update_acknowledgements_total", float64(stats.ValidStatusUpdateAcknowledgements), "Total of valid status update acknowledgements")
+	e.counter(metricsChan, labels, labelValues, "mesos_master_valid_status_updates_total", float64(stats.ValidStatusUpdates), "Total of valid status updates")
+
+	e.counter(metricsChan, labels, labelValues, "mesos_master_registrar_queued_operations_total", float64(stats.RegistrarQueuedOperations), "Total of queued operations in registrar")
```
The description reads as a though this was the number of operations currently queued

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,251 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.gauge(metricsChan, labels, labelValues, "mesos_master_registrar_state_store_duration_milliseconds_max", stats.RegistrarStateStoreMilliSecsMax, "Maximum duration of state store by registrar (in milliseconds)")
+	e.gauge(metricsChan, labels, labelValues, "mesos_master_registrar_state_store_duration_milliseconds_min", stats.RegistrarStateStoreMilliSecsMin, "Minimum duration of state store by registrar (in milliseconds)")
+	summaryQuantileLabels := append(labels, "quantile")
+	e.gauge(metricsChan, summaryQuantileLabels, append(labelValues, "0.5"), "mesos_master_registrar_state_store_duration_milliseconds", stats.RegistrarStateStoreMilliSecsP50, "Duration of state store by registrar (in milliseconds)")
```
Over what time period is this quantile?

If you've going to export quantiles you should put them in a Summary rather than using Gauges. I'd suggest dropping the quantiles.

pvdissel replied:
I will check out the Summary type

brian-brazil commented on main.go:
```diff
@@ -126,73 +109,251 @@ func (e *periodicExporter) Collect(ch chan<- prometheus.Metric) {
+
+	e.counter(metricsChan, labels, labelValues, "mesos_master_registrar_queued_operations_total", float64(stats.RegistrarQueuedOperations), "Total of queued operations in registrar")
+	e.gauge(metricsChan, labels, labelValues, "mesos_master_registrar_registry_size_bytes", float64(stats.RegistrarSizeBytes), "Registrar registry size in bytes")
+	e.gauge(metricsChan, labels, labelValues, "mesos_master_registrar_state_fetch_duration_milliseconds", stats.RegistrarStateFetchMilliSecs, "Duration of state fetch by registrar (in milliseconds)")
```
What exactly is this? I'd be looking for a counter of total time spent

brian-brazil commented:
How are you distinguishing the master and slaves at the label level? What you really want is two scrape configs so that they end up with different job labels.

fabxc commented on main.go:
```diff
@@ -14,50 +14,21 @@ import (
-	localURL       = flag.String("exporter.local-url", "http://127.0.0.1:5051", "URL to the local Mesos slave")
-	masterURL      = flag.String("exporter.discovery.master-url", "http://mesos-master.example.com:5050", "Mesos master URL")
-	metricsPath    = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics")
+	addr = flag.String("web.listen-address", ":9105", "Address to listen on for web interface and telemetry")
```
`go unfmt`? :)

pvdissel replied:
ah, good one. Looks like the IntelliJ Golang plugin (https://github.com/go-lang-plugin-org/go-lang-idea-plugin) doesn't use `go fmt` yet. I shall start using `go fmt` :smile:

juliusv commented on README.md:
```diff
@@ -2,6 +2,17 @@
+    - `/metrics/snapshot`
+    - `/monitor/statistics.json`
+
+For system metrics you can use the [node_exporter](https://github.com/prometheus/node_exporter)
```
Missing period after sentence.

daya commented:
Hi, I am wondering what's preventing this pull request from being merged ? Is it just merge conflicts?

fabxc commented:
There a several open comments that have to be addressed.

daya commented:
@fabxc while we are waiting for the merge, is there a way to configure the exporter to scrape both master and slave metrics from `/metrics/snapshot` ?

discordianfish commented:
Hi everyone. Some time ago I started working on a new mesos-exporter which works for slave and master exposition and has support for the snapshot labels (although lot of those aren't prometheus idiomatic so they might change in the future). I want to replace this mesos_exporter by the other one soon as well (about to kick off a discussion about that). Just so you know before spending time fixing the open comments.

pvdissel commented:
Sorry for the long silence, I'm not able to work on this for atleast a couple of more weeks. So maybe just close this one as @discordianfish wants to replace this one anyway. What do you think @discordianfish, best to close this one?

discordianfish commented:
I'd say yes. Check out the exporter I'm working on which has this feature: https://github.com/mesosphere/mesos_exporter
Right now it blindly exposes every metrics, but I'll change that and massage the data to fit better into the prometheus model (like creating labels for things that sum up to totals when aggregated etc). So expect changes in the metric names if you want to use it today.

Still, thanks a lot for the PR. I know it's frustrating if you spend the time on it and it gets replaced.

StephanErb commented:
Looks like no replacement has landed. Could/should this one here be re-opened?
